dsgrid/registry/registry_database.py

```diff
@@ -0,0 +1,416 @@
+import logging
+import re
+from pathlib import Path
+
+from arango import ArangoClient
+
+from dsgrid.data_models import DSGBaseModel
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.exceptions import DSGValueNotRegistered, DSGInvalidParameter
+from dsgrid.registry.common import Collection, RegistrationModel, Edge
+
+
+GRAPH = "registry"
+MAX_REGISTRY_VERSIONS = 1000
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseConnection(DSGBaseModel):
+    """Input information to connect to a registry database"""
+
+    database: str = "dsgrid"
+    hostname: str = "localhost"
+    port: int = 8529
+    username: str = "root"
+    password: str = "openSesame"
+
+    @classmethod
+    def from_url(cls, url, **kwargs):
+        """Create a connection from a URL."""
+        regex = re.compile(r"http://(.*):(\d+)")
+        match = regex.search(url)
+        if match is None:
+            raise DSGInvalidParameter(f"Invalid URL format: {url}")
+        hostname = match.group(1)
+        port = match.group(2)
+        return cls(hostname=hostname, port=port, **kwargs)
+
+    def url(self):
+        return f"http://{self.hostname}:{self.port}"
+
+
+class RegistryDatabase:
+    """Database containing a dsgrid registry"""
+
+    def __init__(self, client):
+        self._client = client
+        self._graph = client.graph(GRAPH)
+
+    @classmethod
+    def create(cls, conn: DatabaseConnection, data_path: Path):
+        """Create a database to store a dsgrid registry."""
+        client = ArangoClient(hosts=f"http://{conn.hostname}:{conn.port}")
+        sys_db = client.db("_system", username=conn.username, password=conn.password)
+        if not sys_db.create_database(conn.database):
+            raise Exception(f"Failed to create the database {conn.database}")
```

Comment:
I don't feel strongly about this - but here was my take. This should be an impossible situation. We successfully connected to the "system" database with the user's credentials. Next, we tried to create a database but Arango returned False. (Usually, they raise exceptions.) So, this is like an assert but more fool-proof because asserts can be disabled. I never want this error caught.